### ISSUE TITLE
django-no-csrf-token: false positive when form body contains nested {% if %} with a multi-line element

### ISSUE THREAD

**Issue by mulka** [returntocorp/sgrep-rules#3816]
**Describe the bug**

The rule [`python.django.security.django-no-csrf-token.django-no-csrf-token`](https://semgrep.dev/r/python.django.security.django-no-csrf-token.django-no-csrf-token) fires on Django templates that **do** contain `{% csrf_token %}`, when the form body contains a nested `{% if %}` wrapping a multi-line HTML element with a `{{ … }}` interpolation. Collapsing the same element onto one line makes the finding go away.

The rule's main pattern `<form...>...</form>` matches correctly, but the `pattern-not-inside: "<form...>...{% csrf_token %}...</form>"` exemption fails to match through the form body in the `generic` language mode, so the finding is not suppressed.

**To Reproduce**

1. Create `fn.html` (negative control — correctly produces 0 findings):

   ```html
   <form method="post">
       {% csrf_token %}
       <div>
           {% if x %}
               <div>{{ y }}</div>
           {% endif %}
       </div>
       <div>
           <button type="submit">Go</button>
       </div>
   </form>
   ```

2. Create `fp.html` (false positive — rule fires even though `{% csrf_token %}` is present):

   ```html
   <form method="post">
       {% csrf_token %}
       <div>
           {% if x %}
               <div>
                   {{ y }}
               </div>
           {% endif %}
       </div>
       <div>
           <button type="submit">Go</button>
       </div>
   </form>
   ```

3. `diff fn.html fp.html` — the only change is whether one nested `<div>{{ y }}</div>` is on one line or three lines:

   ```diff
   -            <div>{{ y }}</div>
   +            <div>
   +                {{ y }}
   +            </div>
   ```

4. Run:

   ```
   semgrep scan --config p/django --metrics=off fn.html fp.html
   ```

5. Observe that `fn.html` produces 0 findings (correct) and `fp.html` produces 1 finding of `python.django.security.django-no-csrf-token.django-no-csrf-token` pointing at the `{% csrf_token %}` line — even though the CSRF token is clearly present in both files.

**Expected behavior**

Both files contain `{% csrf_token %}` inside `<form method="post">`, so both should produce 0 findings:

```
Ran 1 rule on 2 files: 0 findings.
```

**Actual behavior** (Semgrep 1.160.0, `p/django` latest as of 2026-04-21):

```
fn.html -> 0 findings  (correct)
fp.html -> 1 finding   (FALSE POSITIVE)
    python.django.security.django-no-csrf-token.django-no-csrf-token
```

**Priority**
How important is this to you?
- [ ] P0: blocking me from making progress
- [ ] P1: this will block me in the near future
- [x] P2: annoying but not blocking me

**Screenshots**

N/A — output is a CLI scan, inlined above.

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04
 - Browser: N/A (CLI)
 - Version: Semgrep 1.160.0, running as `semgrep/semgrep:1.160.0` Docker image; rule pack `p/django` pulled from the Semgrep Registry on 2026-04-21

**Smartphone (please complete the following information):**
 - N/A

**Additional context**

**Rule (for reference)**, from [`semgrep-rules/python/django/security/django-no-csrf-token.yaml`](https://github.com/semgrep/semgrep-rules/blob/develop/python/django/security/django-no-csrf-token.yaml):

```yaml
rules:
  - id: django-no-csrf-token
    patterns:
      - pattern: "<form...>...</form>"
      - pattern-either:
          - pattern: '<form ... method="$METHOD" ...>...</form>'
          - pattern: "<form ... method='$METHOD' ...>...</form>"
          - pattern: "<form ... method=$METHOD ...>...</form>"
      - metavariable-regex:
          metavariable: $METHOD
          regex: (?i)(post|put|delete|patch)
      - pattern-not-inside: "<form...>...{% csrf_token %}...</form>"
      - pattern-not-inside: "<form...>...{{ $VAR.csrf_token }}...</form>"
    languages: [generic]
```

**Bisection: required ingredients for the FP**

All four of the following had to be present inside the form body for the FP to fire. Removing any one of them makes the rule correctly recognize the `{% csrf_token %}`:

1. A wrapper `<div>` containing a nested `{% if %}…{% endif %}` block.
2. A **multi-line** HTML element inside that `{% if %}` (opening tag, content, and closing tag on separate lines).
3. A `{{ … }}` interpolation inside that multi-line element.
4. A sibling block after the wrapper `<div>` (e.g. another `<div>` containing a submit button).

| Change to `fp.html`                                                    | Finding? |
| ---------------------------------------------------------------------- | -------- |
| none (as shipped)                                                      | **YES**  |
| collapse inner `<div>...</div>` onto one line                          | no       |
| remove the `{% if %}...{% endif %}` wrapper                            | no       |
| remove the `{{ y }}` interpolation                                     | no       |
| remove the sibling `<div><button>` block                               | no       |
| replace the inner multi-line `<div>` with `<span>` of the same shape   | no       |

**Impact**

The FP appears on realistic Bootstrap-style Django form markup. Real-world example triggering this on a production Django project I maintain:

```html
<form method="post" action="{% url 'signup' %}">
    {% csrf_token %}
    <div class="form-group mb-3">
        {{ form.email }}
        {% if form.email.errors %}
            <div class="invalid-feedback d-block">
                {{ form.email.errors }}
            </div>
        {% endif %}
    </div>
    <div class="d-grid">
        <button type="submit" class="btn btn-primary">Submit</button>
    </div>
</form>
```

The CSRF token is clearly present, but the rule fires anyway, forcing developers to add `nosemgrep` suppressions to every affected form. In the project where I hit this, 4 out of 4 forms flagged by the rule were false positives with this exact shape.

**Hypothesized root cause**

The rule uses `languages: [generic]`, which is Semgrep's token-based ellipsis matcher rather than an HTML/DTL parser. The main pattern `<form...>...</form>` matches fine, but the `pattern-not-inside: "<form...>...{% csrf_token %}...</form>"` exemption appears to fail to match through the form body when it contains the combination above — likely because the generic `...` ellipsis hits its span cap or aligns to the wrong closing token before reaching `</form>`.

Possible fixes:

- Raise `generic_ellipsis_max_span` on the exemption patterns.
- Rewrite the exemption as a simpler check that just looks for the literal substring `{% csrf_token %}` somewhere between the matched `<form>` and `</form>` (e.g. `metavariable-pattern` on the form body, or `pattern-regex`).
- Real DTL parsing support (semgrep/semgrep#182).
